juju/client/facade.py

```diff
@@ -479,37 +479,48 @@ def ReturnMapping(cls):  # noqa: N802
     def decorator(f):
         @functools.wraps(f)
         async def wrapper(*args, **kwargs):
-            nonlocal cls
             reply = await f(*args, **kwargs)
-            if cls is None:
-                return reply
-            if "error" in reply:
-                cls = CLASSES["Error"]
-            if typing_inspect.is_generic_type(cls) and issubclass(
-                typing_inspect.get_origin(cls), Sequence
-            ):
-                parameters = typing_inspect.get_parameters(cls)
-                result = []
-                item_cls = parameters[0]
-                for item in reply:
-                    result.append(item_cls.from_json(item))
-                    """
-                    if 'error' in item:
-                        cls = CLASSES['Error']
-                    else:
-                        cls = item_cls
-                    result.append(cls.from_json(item))
-                    """
-            else:
-                result = cls.from_json(reply["response"])
-
-            return result
+            return _convert_response(reply, cls=cls)
 
         return wrapper
 
     return decorator
 
 
+@overload
+def _convert_response(response: dict[str, Any], *, cls: type[SomeType]) -> SomeType: ...
+
+
+@overload
+def _convert_response(response: dict[str, Any], *, cls: None) -> dict[str, Any]: ...
+
+
+def _convert_response(response: dict[str, Any], *, cls: type[Type] | None) -> Any:
```

Comment:
I see why you split this out, very helpful for testing. I think the overload isn't quite correct according to the implementation, since we can get an error type depending on the response contents, and the code will return a list if the `cls` was defined by a schema with `"type": "array"`.
```py
if TYPE_CHECKING:
    from juju.client._definitions import Error
@overload
def _convert_response(
    response: dict[str, Any], *, cls: type[SomeType]
) -> SomeType | list[SomeType] | Error: ...
@overload
def _convert_response(response: dict[str, Any], *, cls: None) -> dict[str, Any]: ...
def _convert_response(response: dict[str, Any], *, cls: type[Type] | None) -> Any:
```
I think `Any` is fine as the return type of the concrete implementation to avoid touching the implementation itself too much. Could change `dict[str, Any]` to `dict[str, JSONData]` but not a big deal.

Reply:
Let me think about that...
I thought that if there was an error, it would get raised, rather than returned.
But maybe I misread the code.

Reply:
There's something like
```py
if 'error' in reply:
    cls = CLASSES['Error']
... # later
result = cls.from_json(reply['response'])
return result
```
And `Error` is a `Type` subclass in `_definitions.py` -- I don't think it raises on being instantiated or anything unless the `response` is messed up

Reply:
I've looked through the code here and in Connection.rpc() and I'm pretty sure that an "error" can't happen here.
I've added comments to that end.